Convert tactics chapter to prodn, part 2 (#16659)
coq/coq: doc/sphinx/proof-engine/tactics.rst

      for which the tactic will create existential variables.  This lets you
      avoid specifying the asserted statement completely until you start to use
      the hypothesis or later start to prove the statement.

.. tacn:: cut @one_type

   Implements the non-dependent case of the :ref:`App <app_rule>` typing rule,
   the Modus Ponens inference rule.  It is equivalent to
   :n:`enough (@ident: @one_type). revert @ident.`
   This tactic is generally considered obsolete but it is still widely
   used in old scripts.

.. tacn:: pose proof @term {? @as_ipat }
          pose proof ( @ident := @term )
   :name: pose proof; _

   The first form behaves like :n:`assert @one_type {? @as_ipat } by exact @term`
   where :token:`one_type` is the type of :token:`term`.

   .. Théo notes it's odd that the first form uses @term instead of @one_term

   The second form is equivalent to :n:`assert (@ident := @term)`.

   .. tacn:: epose proof @term {? @as_ipat }
             epose proof ( @ident := @term )
      :name: epose proof; _

      While :tacn:`pose proof` expects that no existential variables are generated by
      the tactic, :tacn:`epose proof` removes this constraint.

> .. tacn:: specialize @one_term_with_bindings {? @as_ipat }

Author: Let's come up with examples before I try to reword the description.

Example 1:
```
Goal forall x: nat, (forall n, n >= 0) -> True.
intros.  (* H is forall n : nat, n >= 0 *)
specialize (H x). (* H is x >= 0 *)
```

2. In `List.v`, `remove_length_lt` (line 764) add ". " after the `specialize`, can you explain how the `_` in `specialize (IHl _ Hin)` is used, what it means?

![image](https://user-images.githubusercontent.com/1253341/195702708-66d67ad1-bcc3-4af1-aa04-3470f652bc14.png)

Are there other interesting examples?

Reviewer: > 2\. In `List.v`, `remove_length_lt` (line 764) add ". " after the `specialize`, can you explain how the `_` in `specialize (IHl _ Hin)` is used, what it means?

This is simply a case of a hole being left by the user because it can be inferred by Coq. (The variable `x0` appears in the premise, so by providing `Hin`, Coq is able to infer the value of `x0`.) This is also a common pattern when passing arguments to a lemma in a call to `apply`, etc. Nothing particularly specific to `specialize` here.

> Are there other interesting examples?

I would say that this is representative of the most interesting use case for `specialize` but there might be some edge cases that are not covered. One such edge case is when you call `specialize` on something that is a lemma and not a hypothesis from the local context. In this case, if you provide the `as` clause you can name (or further transform) the result and add it to the local context, but if you don't, then the "specialized" lemma is added as a premise to the goal:

```coq
Require Import Arith.

Goal True.
specialize (Nat.add_comm 1 0).
(* Goal is now: 1 + 0 = 0 + 1 -> True *)
```

Author: How do you know when `specialize` will modify the goal (as in your example immediately above) versus modifying a hypothesis?

And it appears that what you see afterwards is the type of the provided term, not that term itself.

Reviewer: > How do you know when `specialize` will modify the goal (as in your example immediately above) versus modifying a hypothesis?

It modifies the goal (adds a premise to the goal) only if the head of the term passed as an argument is not the name of a hypothesis from the local context *and* there is no `as` clause.

> And it appears that what you see afterwards is the type of the provided term, not that term itself.

That's correct. `specialize (some_lemma args) as H` is the same thing as `pose proof (some_lemma args) as H`.

Author: I think it's time to create a section that starts to explain the relationship between terms and their types, e.g. if the type of H is `forall n m : nat, n >= m` then `(H x)` is like a partial evaluation of a function.  But IIUC there isn't actually a known function for H.  `Show Proof` seems uninformative.

Does that make sense?  Where is a good place to put the new section in the document?  The CIC chapter is much too dense; it would not fit there.

A simpler example:
```
Goal forall x: nat, (forall n, n >= 0) -> True.
intros.  (* H is forall n : nat, n >= 0 *)
Show Proof.
(*
  (fun (x : nat) (H : forall n : nat, n >= 0) =>
 ?Goal@{x:=x; H:=H})
*)
specialize (H x). (* H is x >= 0 *)
Show Proof.
(*
  (fun (x : nat) (H : forall n : nat, n >= 0) =>
   let H0 : x >= 0 := H x in ?Goal@{x:=x; H:=H0})
*)
```

Author: Formally defining "type" conflicts with how it's commonly use in describing computer languages (including OCaml).  Formally the type of `(Nat.add 1 1)` is `forall ...`, which conflicts with the saying that the type of `Nat.add` is a function.  We can only say, so we can say `Nat.add` _is_ a function or `(Nat.add 1) _is_ a function.  Maybe I'll clarify that at some point.

The essential point is that "it's time to create a section that starts to explain the relationship between terms and their types", though probably not today.

Reviewer: > Formally defining "type" conflicts with how it's commonly use in describing computer languages (including OCaml).

No, it doesn't. This is the same notion, just more powerful types are allowed in Coq as they can refer to terms.

> Formally the type of `(Nat.add 1 1)` is `forall ...`, which conflicts with the saying that the type of `Nat.add` is a function.

That isn't correct. The type of `(Nat.add 1 1)` is `nat`, as reported by `Check`. `Nat.add` is a function (not even a dependent function), it has type `nat -> nat -> nat`.

> We can only say `Nat.add` _is_ a function or `(Nat.add 1)` _is_ a function, dropping back to syntax.

I don't understand what you mean by dropping back to syntax here, but for sure, `Nat.add` is a function and it still is a function when it is partially applied to a single argument.

> The essential point is that "it's time to create a section that starts to explain the relationship between terms and their types", though probably not today.

Actually there is one, but it is unfortunately the most complex one, so I've been reluctant to work on it. Still, it should be modified instead of a new section created. It's the "Typing rules" section (a.k.a., the CIC chapter): https://coq.inria.fr/refman/language/cic.html

Reviewer 2: > Formally the type of `(Nat.add 1 1)` is `forall ...`,

I don't understand.  The `Check (Nat.add 1 1)` says `nat`, not `forall ...` for the type.

> the saying that the type of `Nat.add` is a function

We don't say "the type of `Nat.add` is a function" but instead "the type of `Nat.add` is a function type".  (Types are never functions.  But types can be "function types", i.e., "types whose terms are functions".  Perhaps this is confusing terminology, but I believe it is standard.  (This is standard in colloquial English too.  If you ask about the "racing section" of a car dealership, nobody will think that you're asking for the section that participates in races, but will know that you mean for "racing" to apply to the cars in the section.  Similarly if you ask for the "on sale aisle" of a grocery store.  The aisle itself is not on sale.  Similarly, since types can never *be* functions, we infer that "function" must apply to terms of that type.))

Reviewer 3: > saying that the type of `Nat.add` is a function.

Nobody ever says that. I insist: nodody says that. The TYPE of Nat.add IS NOT a function. On the other hand Nat.add ITSELF is a function.

Author: OK, I got the types wrong.  I want to avoid diving deeply into describing these details in this PR, which I think will take time to get right.  The original PR was submitted 14 months ago, parts of which were split off and merged a while ago.

> > The essential point is that "it's time to create a section that starts to explain the relationship between terms and their types", though probably not today.

> Actually there is one, but it is unfortunately the most complex one, so I've been reluctant to work on it. Still, it should be modified instead of a new section created. It's the "Typing rules" section (a.k.a., the CIC chapter): https://coq.inria.fr/refman/language/cic.html

That's what I intended.  (Perhaps I need to think more clearly so you can more easily read my mind :-)  "section" can include "subsection", "subsubsection", ....  As a first step, that chapter needs an introduction that makes sense to the non-expert.  That's going to take a focused effort and is beyond the scope of this PR.

> > saying that the type of `Nat.add` is a function.

> Nobody ever says that. I insist: nobody says that. The TYPE of Nat.add IS NOT a function. On the other hand Nat.add ITSELF is a function.

Consider someone with experience programming in OCaml who is learning Coq on their own and is not familiar with Coq typing.  They may carry over ideas from programming that are incorrect for Coq.  The doc should dispel some of the more likely potential misconceptions.  People do refer to "function types" in OCaml, eg [here](https://www2.lib.uchicago.edu/keith/ocaml-class/data.html#function):

> **Functions**

> In any functional language, one of the most important types of data is the function. Functions can be stored in data structures, passed to functions and returned as the result of functions, and computed dynamically.

Programmers say informal things like "what is the type of `v`?  It's a function".  We need to re-educate people a bit to distinguish between what something _is_ and its type.  As I said "We can only say `Nat.add` _is_ a function or `(Nat.add 1) _is_ a function".

Let's not spend a lot more time on this right now.  I think it's not needed in this PR.

Reviewer 3: > Programmers say informal things like "what is the type of `v`? It's a function".

I think people saying that know perfectly well the abuse here. In ocaml a function is a value and a type cannot be a value so the two worlds are completely separated.

> We need to re-educate people a bit to distinguish between what something _is_ and its type. As I said "We can only say `Nat.add` _is_ a function or `(Nat.add 1) _is_ a function".

Note that saying thet `a -> b` is a "function type" is ok because it means "the type of a function".

> Let's not spend a lot more time on this right now. I think it's not needed in this PR.

OK

      While :tacn:`pose proof` expects that no existential variables are generated by
      the tactic, :tacn:`epose proof` removes this constraint.

.. tacn:: specialize @one_term_with_bindings {? @as_ipat }

>    *First*, the tactic generates a modified term:

Reviewer: Your splitting of the description with *First* / *Then* makes a lot of sense and is easy to read, but I think the tactic is now missing a general one-sentence description. Something like:

>Specializes a Coq term (typically a hypothesis or a lemma) by applying it to arguments.

      While :tacn:`pose proof` expects that no existential variables are generated by
      the tactic, :tacn:`epose proof` removes this constraint.

.. tacn:: specialize @one_term_with_bindings {? @as_ipat }

>    If the :term:`head constant` of :n:`@one_term` has the type

Reviewer: ```suggestion
   If the :term:`head constant` of :n:`@one_term` (in :n:`@one_term_with_bindings`) has the type
```

Author: OK, but I think readers would be able to figure out that `one_term_with_bindings` contains a `one_term`.

Reviewer: This is more consistent with your usual practice elsewhere, and I think that consistency is good in such a reference documentation.

   `forall ...`, the tactic replaces one or more of the

>    quantified variables in the type with terms provided by
>    :n:`@one_term_with_bindings`, either in the form of a

Reviewer: Suggestion to attempt to clarify that it is not the whole `one_term_with_bindings` that we are talking about here (of course, the rest of the sentence and the examples also help):

```suggestion
   quantified variables in the type with arguments provided in
   :n:`@one_term_with_bindings`, either in the form of a
```

Author: I made it "arguments provided by", think that's more natural wording.

Reviewer: OK, but keep in mind that `one_term_with_bindings` provides *both* the main argument to the tactic (the hypothesis or lemma to specialize) and the arguments that are used to specialize it.

   :n:`@one_term_with_bindings`, either in the form of a
   :ref:`function application <function_application>` (which may be partial),
   such as `(H 1)`, or with named binders, such as `H with (n:=1)`.

   If the :term:`head constant` has a non-dependent product type such as
   `A -> B -> C`, the tactic does :term:`forward reasoning`, eliminating one or
   more of the premises.

   Uninstantiated arguments are inferred by unification, if possible, or otherwise
   left quantified in the resulting term.

   If the :term:`head constant` is a hypothesis, the resulting term replaces that
   hypothesis.  Specifying :n:`@as_ipat` will leave the original hypothesis
   unchanged and will introduce new hypotheses as specified by the
   :token:`simple_intropattern`.

   If the head constant is a lemma or theorem, the resulting term
   is added as a new premise of the goal so that the behavior is similar
   to that of :tacn:`generalize`.  In this case, you can use :n:`@as_ipat` to
   immediately introduce the modified term as one or more hypotheses.

   .. exn:: @ident is used in hypothesis @ident.
      :undocumented:

   .. exn:: @ident is used in conclusion.

>       :undocumented:

Author: I think these errors no longer exist.  I can't find "is used in" in a string in the source code.  The variables in the forall are bound in the forall, so they are completely independent.  So they should be removed from the doc.  Agreed?

Reviewer: Can we provide these examples of `specialize` failing (or at the very least update the documented error messages to match the current version)?

Author: I already updated the messages but didn't push them.

I think examples of the error messages are overkill.  I added this: ```If :n:`H` appears in the conclusion or another hypothesis, you must use :n:`@as_ipat` to give a fresh hypothesis.```

      While :tacn:`pose proof` expects that no existential variables are generated by
      the tactic, :tacn:`epose proof` removes this constraint.

.. tacn:: specialize @one_term_with_bindings {? @as_ipat }

>    .. todo PR: Where is @ident now?  Need to adapt

Reviewer: `@ident` is the head constant of the `@one_term` provided as argument.

Author: OK.  Still quite murky on the high level description of specialize, need to see more examples.  The existing text says "This tactic works on local hypothesis :n:`@ident`."  That seems wrong; it need not be a hypothesis and it makes more sense to talk about the `@one_term` than the head constant.

Suppose the `@one_term` is:

![image](https://user-images.githubusercontent.com/1253341/201776143-c27e0906-a9d7-4a63-a2e2-bfeb65eb4dba.png)

The current description says: "The premises of this hypothesis (either universal quantifications or
non-dependent implications) are instantiated by concrete _terms_ from
:n:`@one_term_with_bindings`."  The premises in this case are the `t__i`?  And for `A -> B -> C`, they are A and B?  How do you get multiple _terms_?

   This tactic works on local hypothesis :n:`@ident`. The
   premises of this hypothesis (either universal quantifications or
   non-dependent implications) are instantiated by concrete terms from
   :n:`@one_term_with_bindings`.

   .. todo PR:

>       1) what is "non-dependent implication" above?

Reviewer: "non-dependent implication" would be the same as "non-dependent arrow" or "non-dependent product", i.e. `A -> B` / `forall _ : A, B`. The parentheses just indicates that all premises (both dependent and non-dependent premises) are considered.

Author: I used the equivalent term `non-dependent premise` here.  Right?  Already in the glossary.

      2) In this little example, what exactly is (H x)? It looks almost like
         a function but H doesn't look like a function.  Is this just something
         wierd about specialize or have I missed some basic point?

      Goal forall x: nat, (forall n, n >= 0) -> True.
      intros.  (* H is forall n : nat, n >= 0 *)

>       specialize (H x). (* H is x >= 0 *)

Reviewer: You have missed some basic point. In this context, `H` is a hypothesis of `forall n : nat, n >= 0`, therefore it *is* a function that for any `n` of type `nat` returns a proof that `n >= 0`. Therefore, it can be applied to `x` (which is of type `nat` in this context).

The only thing that is special about `specialize` here (compared to e.g., `pose proof`), is that if you do not provide an `as` clause, then Coq will reuse the name `H` for the specialized hypothesis (exactly as if `specialize (H x) as H` or `pose proof (H x) as H`).

Author: I expect a function would look something like `(fun (x : nat) (H : forall n : nat, n >= 0) => ?Goal)` rather than a `forall`.  Is it that H really is a function but the UI only shows the type of H?  Or is H exactly the displayed type, which gets mapped/converted back into a function?  Whichever it is, I don't think that is covered anywhere in the doc.

Reviewer 3: Coq prints : `H : forall...`, which definitely means that the *type* of H is `forall...`. Otherwise you would have `H := forall...`.
However sometimes we say that the hypothesis `H` *is* `forall...` because that is how mathematicians talk about hypothesis, but this is misleading.

Author: Hi Pierre,

Thanks for offering to help.  Let's take things one at a time:

`(H x)` in the specialize tactic on line 1608 looks like the application of a function (including supporting partial application).  If that's a correct way to think about it, I'd like to say something clear and strong about the correspondence between a function and its associated type (a forall).  So far I've not learned how to compute the type of a term--whether it's not in the doc at all or if it's there but not explained in an accessible form (e.g. it's in the rather dense [CIC chapter](https://coq.inria.fr/distrib/current/refman/language/cic.html)).  Fully explaining how to compute types is beyond the scope of this PR.

Or I could just make a weaker statement in this tactic noting the correspondence.  The context shows the type of H.  Every term has a type, but IIUC not every type has a corresponding term.  e.g. if H is not provable, then there is no associated term.  Even if H is provable, Coq probably doesn't know the what the term is (imagine H: <Fermat's Last Theorem>).

Author: I suspect the `(H x)` is a merely a convenient notation since Coq (generally?) doesn't know the specific function whose type is shown for H.

Reviewer 3: Indeed, when `H: T` appears in a proof context, it means that `H` is a *hypothesis* so it does NOT say there is a closed term with type `T`. Instead it says that we must build a proof *under the assumption* that there is a proof of `T`, that we shall name `H`. If the type `T` happens to be a product (dependent or not), then the proof may contain applications of the form `H x`.
I think this is explained in the early chapters explaining the concepts of interactive proofs, we don't really need to come back to this in the chapters on tactics IMHO.

Author: There's something unexplained going on that's not explained in the doc.  I'd like to add the explanation in an appropriate section.  `(f x)` indicates applying a function f to an argument x.  From its type, `H` must be a function, but an unknown one, so how can you apply it in `(H 1)`?  Coq somehow treats the type `forall n m ...` as a function definition, which deserves an explanation.

```
Goal (forall n m, n >= m) -> True.
intros.   (* H : forall n m : nat, n >= m *)
let x := type of (H 1) in idtac x.  (* (forall m : nat, 1 >= m) *)
```

It seems to work the same way for theorems, but they don't look like functions in the `Print`.  That should be explained, too:

```
Theorem F: forall n m, n >= m.
Admitted.
Goal True.
let y := type of (F 1) in idtac y.  (* (forall m : nat, 1 >= m) *)
Print F.
(*
  *** [ F : forall n m : nat, n >= m ]

  Arguments F (_ _)%nat_scope
*)
```

Can you explain these enough so I can describe how they work?

Reviewer 3: I see what you mean but this is old plain higher order programing : for instance, take the `List.map` function of ocaml:

```ocaml
let map (f:'a ->'b) (l:'a list): 'b list =
  match l with
  | [] -> []
  | e::l' -> f e :: map f l'
```

`f` is only a variable here and we apply if to `e`, and the type of subterm `f e` is `'b`. Do you really think we need to explain that here? I mean it is really really basic functional programing.

Reviewer 3: When doing interactive proof we are manipulating the typing environment in the same way the typing algorithm does: in the example above when the typing algorithm reaches `f e` there is a typing environment, which includes parameters `f` and `l`, but also pattern variables `e` and `l'` that were added while diving into the term. The difference in interactive proof is that we build the term *while* typing it.
Explaining this as a preamble could be nice for users coming from outside functional programing community. But this should be done as separate section or chapter.

Reviewer 3: Illustrating this with the map function defined interactively:
```coq

Definition map {A B:Type}: forall  (f:A -> B) (l:list A) , list B.
  fix map 2.
  intros.
  case l.
  - exact nil.
  - intros e l'.
    exact (cons (f e) (map f l')).
Defined.
```

Reviewer 3: I hope this example also shows that there is absolutely no difference between applying a theorem, a function or a hypothesis. There is a difference in the ways these will be used afterward though: lemmas and hypothesis are rarely (never for hypothesis) used for computing, whereas functions may be called and reduced on concrete parameters.

Reviewer 3: Also: `A -> B` is *really* just a syntactic suger for `forall x:A, B` when `B` does not mention `x`. Just type `Check (forall x:nat, bool)` in Coq to see that coq prints: `nat -> bool: Set`. This is explained [very early](https://coq.inria.fr/distrib/current/refman/language/core/assumptions.html?highlight=forall#functions-fun-and-function-types-forall) in the documentation.

Author: I'm well aware of these:

> `f` is only a variable here and we apply if to `e`, and the type of subterm `f e` is `'b`

> `A -> B` is *really* just a syntactic sugar for `forall x:A, B` when `B` does not mention `x`. 

This suggests my point:

> ... we build the term *while* typing it.

The evaluation process is not fully described.  H (a hypothesis) and F (a theorem) are handled differently from, say Nat.add (a fixpoint).  I think it works like this: For `(H 1)`, the type of a hypothesis H is known but its value is never known/generated/used by Coq.  Since the type of H is a dependent product, Coq infers the new type (removing the first dependent premise `n` in the `forall` ).  `let x := type of (H 1) in ...` is valid because the type is known, but `let x := (H 1) in ...` is not because the value is unknown.

Even though the value (proof) of a proven theorem is known, it appears that `let x := (F 1) in ...` is never valid, probably because it's not a useful thing to do.  Right?

In contrast, `let x := (Nat.add 1) in ...` gives you a new function.

Can you always get a value for any variable that isn't a hypothesis or a theorem?

> Explaining this as a preamble could be nice for users coming from outside functional programing community.

I think "functional programming" has no fixed meaning; there's a set of features that are associated with the term, but each language has its own definition of the term.  So good to explain.

If this seems on-track, I'll draft a proper description to add to the doc.

Author: Also, if the hypothesis only has nondependent premises, then `(H <something>)` is never valid?  E.g. something like this won't work even with tweaks:

```
Goal (0=0 -> 1=1) -> False.
intros.    (* H : 0 = 0 -> 1 = 1 *)
let x := type of (H (0=0)) in idtac x.
(*
  In environment
  H : 0 = 0 -> 1 = 1
  The term "0 = 0" has type "Prop"
  while it is expected to have type "0 = 0".
*)
```

Reviewer 3: Sorry but I think there is an important aspect of coq that you misunderstand.

> The evaluation process is not fully described.

Because it does not need to be described: we build a *proof*, i.e. a term having a certain type, we don't compute with it. *After* the term is built, we may try to compute with it but it is another story.

> H (a hypothesis) and F (a theorem) are handled differently from, say, Nat.add (a fixpoint).

As far as typing is concerned: NO DIFFERENCE except for conversion: `H` does not unfold whereas theorems (even if opaque), whereas `F` and `Nat.add` do. So the only difference is between `H` and `F`, or between `H` and `Nat.add`.

> I think it works like this: For (H 1), the type of a H is known but its value is never known/generated/used by Coq.

`H` is its own value, that's all. It *is* used, as a variable.

> Since the type of H is a dependent product, Coq infers the new type (removing the first dependent premise n in the forall ).

? This has nothing to do with the product being dependent or not. In both cases `H` needs something having the *type* appearing in the left member of the product.

> let x := type of (H 1) in ... is valid because the type is known,

Yes, but are you aware that `x` has type `Prop` here?  And are you aware that the `let` here is at Ltac level here ? 

```coq
Goal (forall n:nat, n=n) -> 1=1.
intros H. 
let x := type of (H 1) in
idtac x. (* 1 = 1*)
let x := type of (H 1) in
let t := type of x in
idtac t. (* Prop *)
```

> but let x := (H 1) in ... is not because the value is unknown.

?? What? Of course it is valid, but since you are using Ltac-let you need to use `constr:` in front of the term otherwise `(H 1)` is taken as an tactic application. But it is perfectly valid, and `x` is "something of type `(H 1)`":

```coq
Goal (forall n:nat, n=n) -> 1=1.
intros H. 
let h := constr:(H 1) in exact h.
```
Using the constr-let:
```coq
Goal (forall n:nat, n=n) -> 1=1.
intros H. 
exact (let x := H 1 in x).
```

> Even though the value (proof) of a proven theorem is known, it appears that let x := (F 1) in ... is never valid, probably because it's not a useful thing to do. Right?

Wrong, very wrong. See examples above and below.

Some "names" have a _definition associated_ with them, while others don't. There are treated exactly the same by the typer except conversion. Period. Don't overthink it, there is no trick here.

`let x := t in u`, means that in `u` the _definition associated_ with name `x` is `t`. Period. No trick. Here `t` is `(F 1)`, so be it, `(F 1)` is a perfectly valid term in an environment where `F` is of type `(nat -> something)`.

> In contrast, let x := (Nat.add 1) in ... gives you a new function.

> Can you always get a value for (V value) when V doesn't refer to a hypothesis or a theorem?

Define "value" please: Since everything is a term in coq (types are terms too), "value" must be some kind of term:

- any term? `(H 1)` is a term.
- a term that is not a type? `(H 1)` is a type.
- a term in normal form? `(H (S 0))` is a term in normal form.
- a ground term (no more names in it) ? `(H 1)` is not ground.

>> Explaining this as a preamble could be nice for users coming from outside functional programing community.

> I think "functional programming" has no fixed meaning; there's a set of features that are associated with the term, but each language has its own definition of the term. Coq, OCaml, Java all considered functional, but different in many ways. So good to explain.

Right. I should have said regular typed lambda-calculus.

Actually I am starting to think you are indeed taking coq for a programming language: in ocaml `42` is meant to represent a binary integer in computer's memory once compiled and executed. In Coq (and lambda-calculus) it is **not**. It represents only itself: a term of the form `S ... (S 0)...)`. Nothing else.

> If this seems on-track, I'll draft a proper description to add to the doc.

It is not. What you are trying to explain does not exists afaiu.

> Also, if the hypothesis only has nondependent premises, then (H <something>) is never valid? E.g. something like this won't work even with tweaks:

> Goal (0=0 -> 1=1) -> False. intros. (* H : 0 = 0 -> 1 = 1 *) let x := type of (H (0=0)) in idtac x. (* In environment H : 0 = 0 -> 1 = 1 The term "0 = 0" has type "Prop" while it is expected to have type "0 = 0". *)

??? this is a standard type error, don't over-interpret it, you just made a mistake: The argument of `H` must be of TYPE `(0=0)`, not be `(0=0)` itself. If you do `H (eq_refl 0)` it is perfectly valid (and may be useful of course).

```coq
Goal (0=0 -> 1=1) -> False.
intros.    (* H : 0 = 0 -> 1 = 1 *)
let x := type of (H (eq_refl 0)) in idtac x.
```

I am sorry but I think you should not try to write in this section until you understand what you are getting wrong here.

Author: > I am sorry but I think you should not try to write in this section until you understand what you are getting wrong here.

That's the point of our discussion.  I'm asking about points that aren't well covered in the documentation.  I iteratively guess (often imperfectly) how things work and eventually come up with an accurate description.

Let's focus on this one point:

> > I think it works like this: For (H 1), the type of a H is known but its value is never known/generated/used by Coq.

> `H` is its own value, that's all. It *is* used, as a variable.

That doesn't make sense to me.  Consider:

```
Goal (forall n m, n >= m) -> True.
intros.  (* H : forall n m : nat, n >= m *)
let x := type of (H 1) in idtac x.
```

`(H 1)` represents the application of a function, however you seem to say that the value of `H` is `forall n m : nat, n >= m`, which is not a function AFAICT.  How does that make any sense?

--------------
Let me respond briefly to a few other comments (and come back to others a bit later):

> > The evaluation process is not fully described.

> Because it does not need to be described: we build a *proof*, i.e. a term having a certain type, we don't compute with it.

By "evaluation process" I meant whatever processing Coq does with a term, such as to check its validity.  I was not referring specifically to evaluating function applications as computations.

> Actually I am starting to think you are indeed taking coq for a programming language: in ocaml `42` is meant to represent a binary integer in computer's memory once compiled and executed. In Coq (and lambda-calculus) it is **not**. It represents only itself: a term of the form `S ... (S 0)...)`. Nothing else.

Of course not.  Everyone learns the representation of Nat at the outset of learning Coq.

Reviewer 3: > That doesn't make sense to me. Consider:
> 
> ```
> Goal (forall n m, n >= m) -> True.
> intros.  (* H : forall n m : nat, n >= m *)
> let x := type of (H 1) in idtac x.
> ```
> 
> `(H 1)` represents the application of a function

Yes.

> however you seem to say that the value of `H` is `forall n m : nat, n >= m`, 

Again, what makes you think that?? When it is written `H : XXX` it means that `H` HAS TYPE `XXX`. It does not say anything about its value. Here `H`  HAS  TYPE  `forall n m : nat, n >= m`. `H` has no value, only a type: it is *assumed* instead of being `defined`. Whereas *constants* have both a type AND a *definition* (that you may call a value but it is misleading imho). 
The **TYPE** tells us that `H` can be applied to 2 arguments of type `nat`. So we know *by typing* that `H 1` is well typed.

> By "evaluation process" I meant whatever processing Coq does with a term, such as to check its validity. I was not referring specifically to evaluating function applications as computations.

Well, evaluation and typing are two very different things. If you want to write the documentation you should start by never naming one by the other imho. Otherwise you will confuse the reader.

- Checking the validity = typing.
- evaluation = reduction 

> > Actually I am starting to think you are indeed taking coq for a programming language: in ocaml `42` is meant to represent a binary integer in computer's memory once compiled and executed. In Coq (and lambda-calculus) it is **not**. It represents only itself: a term of the form `S ... (S 0)...)`. Nothing else.
> Of course not. Everyone learns the representation of Nat at the outset of learning Coq.

OK but you seem obsessed with "values":

> The evaluation process is not fully described. ...
> I think it works like this: For (H 1), the type of a H is known but its value is never known/generated/used by Coq. ...
> let x := type of (H 1) in ... is valid because the type is known, but let x := (H 1) in ... is not because the value is unknown.

But the whole proof mode is a type-driven term construction tool. It does not care about "values", only types. Coq's language happens to be dependently types so the typer needs to perform conversion from time to time. Which implies to use *reduction*. The reduction is the only moment the typer looks at the definitions associated with names. Otherwise looking at the *type* associated with names is enough, and there is no (*no*) difference netween variables and constants.

Author: The current description of function application is [here](https://coq.inria.fr/distrib/current/refman/language/core/assumptions.html#function-application).  `H` is not a function and the type of H (`forall n m : nat, n >= m`) is not a function, so how can we briefly and concretely explain, somewhere, the interpretation of `(H 1)`?  Should that be in the description of function application or elsewhere?  It probably doesn't belong in the description of `specialize`.  I think this should only take a few sentences.

> Well, evaluation and typing are two very different things. If you want to write the documentation you should start by never naming one by the other imho. Otherwise you will confuse the reader.

Did you mean to be snarky here?  I described exactly what I meant by "evaluation" for the purpose of our discussion, which is standard terminology for software (though not for Coq), which you quoted.

Reviewer 3: This discussion is going nowhere sorry. I probably don't understand what you are trying to say. Last attempt:
`H` is a variable, it takes place of something else. And it can be used as the thing it replaces. Examples:
- In a term like `fun H:nat => H + 1`, `H` is a variable of type `nat` and therefore we can use it as a `nat` (like adding `1` to it).
- In a term like `fun H:(nat -> nat) => (H 1)`, `H` is a variable of type `nat -> nat` and therefore we can use it as a function (since its type is a product), like applying it to `1`.
- In a term like `fun H:(forall n m : nat, n >= m), (H 0 1)`,  `H` is a variable of type `forall n m : nat, n >= m` and therefore we can use it as a function (since its type is a (dependent) product), like applying it to `0` and `1`.

So yes we can say that "`H` is not a function" in the last example as we can say that "`H` is not a natural" in the first example, but it is a variable with a functional type so yes of course we can use it as a function. That's precisely what variables are meant to be used, I really don't see your point here.

Author: Thanks for trying to explain.  Some discussions don't work very well by email.  Perhaps I can figure this out with a little help from @Zimmi48 when he has some time.

A last (full) restatement of my question:

Consider

```
Goal (forall n m, n >= m) -> True.
intros.                             (* H : forall n m : nat, n >= m *)
let x := type of (H 1) in idtac x.  (* (forall m : nat, 1 >= m) *)
```
 
`(H 1)` represents the application of a function.  However `H` is not a function, merely a variable with a known type. The type of `H` (`forall n m : nat, n >= m`), which doesn't look like a function, either.  So how can this term be considered application of a function?  The doc for function application [here](https://coq.inria.fr/distrib/current/refman/language/core/assumptions.html#function-application) says the first term (`H`) must be a function.  I thought the answer might be something like:
- the description of function application  is incorrect/incomplete, OR
- The processing of `(H 1)` looks at the type of `H` to infer the type of `(H 1)`, which, if correct, ought to be mentioned in the doc.

(In contrast, Nat.add is function (right?), so `(Nat.add 1)` fits the documentation for function application.  So seems like there are 2 cases to consider for function application.)

Reviewer 3: Although the documentation describes how a function is supposed to be built, it does not say that only functions can be applied: syntax of application is [here](https://coq.inria.fr/distrib/current/refman/language/core/assumptions.html#grammar-token-term_application). The form is `term1 arg`, where [term1](https://coq.inria.fr/distrib/current/refman/language/core/basic.html#grammar-token-term1) is any kind of term/evar/etc. So really the only restriction of application is typing.

Author: Well, that doc is under the heading **Function application** and it says things like
![image](https://user-images.githubusercontent.com/1253341/205215583-cdffde48-178e-4ebc-80f2-24c5548c2660.png)
I read that as requiring a function.  But if what you say is correct, I will revise the description.  Could you give an example in which `term1` is not a function?

I expect that `(H 1)` always gives an error if the type of `H` is not a `forall ...`.  Counterexamples welcome.  (I like being proved wrong, it means I'm learning something.)

Reviewer 3: Yes, that is right: the only terms that can be applied are those who have a product type. Whatever the term is (a hypothesis (i.e. a variable), an evar, a complex term, a lemma etc...).
Your initial question was "why can we apply `H` to something since `H` (of type `forall bla`) is a hypothesis and not a function". The answer is: yes it is a function since it has a product type.

Reviewer 3: > ![image](https://user-images.githubusercontent.com/1253341/205215583-cdffde48-178e-4ebc-80f2-24c5548c2660.png)

Remark: I don't understand the meaning of the subscript `fun` in the documentation. It seems to refer to `term1` (when I click on it it takes me to the `term1` entry, so I guess it is just a way to distinguish this term from the other `term1` of the expression. Is this a global convention? Any subscrip added to a grammatical entry is actually just a tag for easy reference?

Author: > Any subscript added to a grammatical entry is actually just a tag for easy reference?

Exactly.

Author: The current text suggests that you can specialize hypotheses such `H : A -> B`, but I wasn't able to construct a working example.  If is indeed possible, could you provide a trivial example?

This is that last thing I tried:

```
Goal forall A B:Prop, (1=1) -> (A -> B) -> True.
intros A B H1 H.
(*
  A, B : Prop
  H1 : 1 = 1
  H : A -> B
*)
specialize (H H1).
(*
  In environment
  A, B : Prop
  H1 : 1 = 1
  H : A -> B
  The term "H1" has type "1 = 1" while it is expected to have type "A".
Raised at file "pretyping/coercion.ml", line 529,
```

Reviewer 3: This is a simple type error:
- `H` is of type `A -> B` so you need to apply it to something of type `A`.
- you apply it to `H1`, which does not have type `A`, type error.
- If you had something of type A, then you could apply `H` to it:

Example:
```coq
Goal forall A B:Prop, (A -> B) -> A -> True.
Proof.
  intros A B H prfA. 
  specialize (H prfA).
```

Author: Thanks.  That seems a more intuitive way to do forward reasoning in hypotheses than `apply`.  `specialize` changes `H` while `apply` changes `prfA`.  Maybe worth mentioning as a tip in the doc.

For example:

```
Goal forall A B:Prop, (A -> B) -> A -> True.
Proof.
  intros A B H prfA.
  apply H in prfA.
(*
  1 subgoal
  A, B : Prop
  H : A -> B
  prfA : B
*)
```

versus

```
1 subgoal
A, B : Prop
H : B
prfA : A
```

`apply` may create new goals (and `specialize` can only apply premises in order):

![image](https://user-images.githubusercontent.com/1253341/205519170-5b5e5fa0-8a8a-417e-83ce-c3d962c75d82.png)

Reviewer 2: > I expect that `(H 1)` always gives an error if the type of `H` is not a `forall ...`. Counterexamples welcome. (I like being proved wrong, it means I'm learning something.)

Not germane to the discussion here, but there are three technical counterexamples: the type of `H` may be reducible to a product without being syntactically a product; the type of `H` may be an evar which then is instantiated with a product during unification; or the type of `H` may be a type with a registered coercion to `Funclass`.

Reviewer 2: Also, is "function" defined anywhere in the documentation?  It seems like a lot of the confusion here is about what, exactly, counts as a "function".  Is it something whose type is `forall ...`?  Is it something which reduces to `fun ...` or `fix ...` or `cofix ...` or ...?  I'd go with the former, even though it takes some adjustment to see `Axiom DNE : forall P, ~~P -> P` as making `DNE` a function.

Author: @JasonGross Regarding your first three comments, I now have "If the head constant of _one_term_ has the type **forall ...**" and then give (H 1) as an example of function application.  I think makes it unnecessary to mention your corner cases here.  Look at the what I wrote (after I push my changes again).

I added a todo about your last comment in `assumptions.rst`.  I want to wrap up the tactics chapter.

.. tacn:: specialize @one_term_with_bindings {? @as_ipat }

   If the :term:`head constant` of :n:`@one_term` has the type
   `forall ...`, the tactic replaces one or more of the
   quantified variables in the type with terms provided by
   :n:`@one_term_with_bindings`, either in the form of a
   :ref:`function application <function_application>` (which may be partial),
   such as `(H 1)`, or with named binders, such as `H with (n:=1)`.

   If the :term:`head constant` has a non-dependent product type such as
   `A -> B -> C`, the tactic does :term:`forward reasoning`, eliminating one or

>    more of the premises.

Reviewer 3: In the dependent case it is also forward reasoning: replacing `H:forall a b:A, (T a b)` by `forall b:A, T 1 b` is also eliminating a premise. I think you can just remove that sentence.

Author: The doc describes forward and backward reasoning [here](https://coq.gitlab.io/-/coq/-/jobs/3454025112/artifacts/_build/default/doc/refman-html/proof-engine/tactics.html#term-forward-reasoning).  The sense for "forward reasoning" described [here](https://en.wikipedia.org/wiki/Forward_chaining) makes sense for `A -> B -> C` but I don't see how it makes sense for dependent products.

Also the second example refers to an `apply` example using forward reasoning in hypotheses, so it made sense to mention the term here.

   Uninstantiated arguments are inferred by unification, if possible, or otherwise

>    left quantified in the resulting term.

Reviewer 3: This is true, but there is a difference between `specialize (X ...)` and `specialize X with ...`.
- in the form `specialize (t1 ... tn)`,  the term `(t1 ... tn)` needs to be well formed and well typed. In particular you cannot omit the ith argument if you want to give the i+1th one. So the only form of partial application is by omitting the *last* arguments of `H`. This is the usual curryfied partial application.
- in the form `specialize t1 with (3:=...) (6:=...)... .` You can give arbitrary arguments to `H`, eliminating for instance the third and sixth ones, and leave the others quantified (actually a few others may be instantiated by side effect). This form is therefore more powerful in the sens that it gives more control on picking the argments you want to instantiate.
Sorry for poor english here. Hope this is clear.

Author: I think being well formed and well typed is almost always required, so I don't want to repeat that here (or 100 other places).  The reader has work a little to learn pervasive facts about Coq.

I changed one of the examples to use a numbered argument, whose description is a click away at `one_term_with_bindings`.

> Sorry for poor english here. Hope this is clear.

More than good enough for our informal conversation and quite clear.

   If the :term:`head constant` is a hypothesis, the resulting term replaces that
   hypothesis.  Specifying :n:`@as_ipat` will leave the original hypothesis
   unchanged and will introduce new hypotheses as specified by the
   :token:`simple_intropattern`.

   If the head constant is a lemma or theorem, the resulting term
   is added as a new premise of the goal so that the behavior is similar
   to that of :tacn:`generalize`.  In this case, you can use :n:`@as_ipat` to
   immediately introduce the modified term as one or more hypotheses.

   .. exn:: @ident is used in hypothesis @ident.
      :undocumented:

   .. exn:: @ident is used in conclusion.
      :undocumented:

   .. example:: partial application in :tacn:`specialize`

      .. coqtop:: reset none

         Goal (forall n m: nat, n + m = m + n) -> True.

      .. coqtop:: out

         intros.

      .. coqtop:: all

>          specialize (H 1). (* equivalent to: specialize H with (n := 1) *)

Reviewer 3: Maybe add an example illustrating my remark above (picking which parameters to instantiate)?

   behavior, such as:

   .. coqdoc::

      Ltac instantiate_ltac_variable ev term :=
        let H := fresh in
        pose ev as H;
        instantiate (1 := term) in (value of H);
        clear H.

   :n:`in @ident`
     Selects the hypothesis :n:`@ident`.

   :n:`in %|- *`
     Selects the goal.  This is the default behavior.

   :n:`in ( type of @ident )`
     Selects existential variables in the type of the
     :term:`local definition <context-local definition>` :n:`@ident`.
     (The body is not included.)

   :n:`in ( value of @ident )`
     Selects existential variables in the body of the
     :term:`local definition <context-local definition>` :n:`@ident`.
     (The type is not included.)

.. tacn:: absurd @one_type

   :n:`@one_type` is any proposition
   :g:`P` of type :g:`Prop`. This tactic applies False elimination, that is it
   deduces the current goal from False, and generates as subgoals :g:`∼P` and
   :g:`P`. It is very useful in proofs by cases, where some cases are
   impossible. In most cases, :g:`P` or :g:`∼P` is one of the hypotheses of the
   local context.

.. tacn:: contradiction {? @one_term_with_bindings }

>    Tries to prove the current goal by finding a contradiction, such as

Author: Odd that the last line fails.  I would have expected it would introduce the negation of the statement.  Or is that unsound in  COC.  Also not sure whether the no-op is a feature or a bug.

```
Goal ~0=0 -> True.
intros.
contradiction H.  (* 0 = 0 *)
contradiction H.  (* no op *)
Abort.
Goal 0 = 0 -> True.
intros.
contradiction H. (* Not a contradiction. *)
```

Reviewer: The no-op is normal. What `contradiction H` does in this case is that it replaces the current goal, whatever it is, with a goal asking you to prove the non-negated version of `H`. This is what happens in your example. But such a behavior is naturally "idempotent" (if you apply it a second time, you get the same result).

The behavior you expected for the last line is very reasonable, and I would have expected it as well, but it turns out that it is not the case (as I described in my second bullet point in https://github.com/coq/coq/pull/16498#discussion_r993126911). IMHO, it would be reasonable to extend `contradiction H` to not fail in this case. OTOH, there is already the `contradict` tactic which seems strictly more powerful for such uses.

Reviewer: I'm reviewing uses of `contradiction` in the standard library. Apparently, it doesn't just look at the local context. This example shows that it can look into goal premises as well:

```coq
Lemma absurd_eq_bool : forall b b':bool, False -> b = b'.
Proof.
  contradiction.
Qed.
```

This is further confirmed by this other example:

```coq
Goal forall A B:Prop, A -> ~A -> B.
contradiction.
Qed.
```

Something else I just noticed is that `contradiction H` is less powerful than `contradiction`. Taking your example

```coq
Goal ~0=0 -> True.
intros.
contradiction H.
(* leaves the goal 0 = 0 which now requires an explicit call to reflexivity *)
Undo.
contradiction.
Qed.
```

Reviewer: The only interesting case of using `contradiction` with an argument in the standard library is when using it with a term as argument (in the case it is used with a hypothesis, it can be replaced with `contradict`):

```coq
Require Import Arith.

Goal forall (A : Prop), 0 < 0 -> A.
intros. (* In this case, the call to intros is necessary *)
contradiction (Nat.lt_irrefl 0). (* Nat.lt_irrefl 0 is a proof of ~(0 < 0) *)
(* No more goals *)
Undo. (* Obviously this is strictly the same as doing: *)
pose proof (Nat.lt_irrefl 0); contradiction. (* No more goals *)
Restart. (* And other solutions are possible *)
intros * ?%Nat.lt_irrefl; contradiction.
Qed.
```

Author: > This example shows that [`contradiction` (no arg)] can look into goal premises as well.

The no argument form (only) does an `intros` first (line 142).  (I wonder a little how well-thought through the behavior is.)

The code is in `contradiction.ml`.  Start at the end of the file.  But hard for me to learn much from the code, maybe it makes more sense to you.  FWIW, we really should have some quality internals documentation for how the tactics code works--it seems pretty impenetrable even though what it does is, in many cases, probably simple.  Very few comments.

Author: > However, it is able to detect when it is solved by `assumption`.

In which example?  It didn't work in the example following your comment.

BTW, this commenting interface is awkward for commenting on longish comments with multiple points.  I've added some "JEF:" text to bottom of your comments.

Author: > So `contradiction H` is not able to detect that the remaining goal is solved by `reflexivity`.

Maybe the no-argument form is applying a reduction to address this?  Perhaps you can pick it out in the source code and tell me what reduction it is?

Reviewer: > In which example? It didn't work in the example following your comment.

The example below the quoted sentence calls `contradiction H1` and `contradiction H2` and the tactic detects that the second hypothesis is need is in the context as well.

   impossible. In most cases, :g:`P` or :g:`∼P` is one of the hypotheses of the
   local context.

.. tacn:: contradiction {? @one_term_with_bindings }

   Tries to prove the current goal by finding a contradiction.  If

>    :n:`@one_term_with_bindings` is not provided, the tactic first does a
>    :tacn:`intros`

Reviewer: ```suggestion
   :n:`@one_term_with_bindings` is not provided, the tactic first does an
   :tacn:`intros`
```

>    then succeeds if there is a hypothesis :n:`False` in the updated context, or

Reviewer: ```suggestion
   then succeeds if there is a hypothesis :n:`False` (or of another empty inductive type) in the updated context, or
```

>    if there is a pair of hypotheses where one is simple negation of the other

Reviewer: Not sure what "simple" brings here:

```suggestion
   if there is a pair of hypotheses where one is the negation of the other
```

>    (i.e. :n:`P` and not :n:`~P`).

Reviewer: Missing in this paragraph is the case where a hypothesis is the negation of "a singleton inductive type (e.g. :g:`True` or :g:`x=x`)"

Author: How is `x=x` a singleton inductive type?

This seems related what I ask about [below](https://github.com/coq/coq/pull/16659#discussion_r1022362456)

Reviewer 2: "singleton inductive" means "inductive type (family) with one constructor".  `eq` has only one constructor, `eq_refl`.  So perhaps it would be more accurate to say "singleton inductive goal which can be solved simply by running `constructor`" or something like that?

Reviewer: I think we should follow Jason's suggestion here.

Author: I think it's not helpful to refer to `constructor`.  Not as easy to understand as what I have.  (Nitpick: in a hypothesis, not in the goal.)

Author: @JasonGross For `contradiction`, IIUC `(1=0)` is an inductive with a single constructor (and a true, provable statement).  Can you suggest better wording that excludes `(1=0)`?  Probably it should say that `T` in `~T` must "reduce" to True (but using more precise language, e.g. exactly what reduction)?

![image](https://user-images.githubusercontent.com/1253341/209253716-fc48f48b-ddaa-477f-82b8-3c05a2af171d.png)

```
Goal ~(1=0) -> False.
intros.
contradiction.  (* No such contradiction *)
```

Reviewer: There is no "reduction" that would transform `1=0` to `True` (or to anything else, really). I think having sufficiently precise language excluding this case was precisely Jason's objective when he wrote "which can be solved by running `constructor`".

Author: > There is no "reduction" that would transform `1=0` to `True` (or to anything else, really).

A misunderstanding.  My question asked how to exclude `(1=0)`. `~(1=0)` doesn't work for `contradiction`.  "Probably it should say that `T` in `~T` must "reduce" to True (but using more precise language, e.g. exactly what reduction)?" is about cases that work, such as `(0=0)`.

How about this?  (The "for which ..." is a bit of novel language, feel free to suggest better wording.  I wanted to indicate that `constructor` is in a different proof):

   - there is a hypothesis that is the negation of a singleton inductive type `T`
     (such as :n:`True` or :n:`x=x`) for which `Goal T. constructor.` is a complete
     proof.

Reviewer: Your proposal is acceptable to me, but I think I have a preference for a slightly simpler form such as:

- there is a hypothesis that is the negation of a singleton inductive type that Coq could prove by running `constructor` (e.g. `True` or `x=x`).

IMHO it kind of makes sense to consider that Coq would create a proof that would at some point run `constructor` internally here, much like we sometimes write that Coq runs `intros` internally.

Anyway, feel free to chose your wording, mine, or something in between.

Author: I think I'll go with this:

   - there is a hypothesis :n:`~P` where `P` is a singleton inductive type
     (such as :n:`True` or :n:`x=x`) that `Goal P. constructor.` can prove.

   Use the :tacn:`discriminate` tactic to prove the current goal when there
   is a hypothesis with an impossible structural equality such as
   :n:`0 = 1`.

>    If :n:`@one_term_with_bindings` has the type :n:`False`, the tactic

Reviewer: ```suggestion
   If :n:`@one_term_with_bindings` is provided and has the type :n:`False` (or another empty inductive type), the tactic
```

>    proves the current goal.  If the type is a negation, e.g. :n:`~P`, then

Author: It looks like the tactic does some reductions (e.g. a cbv?).  Maybe it reduces all the hypotheses before checking for contradictions?  What wording would you suggest?

```
Goal forall (A : Type) (x : A), ~(x = x) -> False.
intros.
(*
1 goal
A : Type@{Top.1}
x : A
H : not (@eq A x x)
______________________________________(1/1)
False
*)
contradiction.
Qed.
```

Reviewer: I have no idea what kind of reduction it does. I think the current description is good enough for now (barring my other comments).

>    the current goal is replaced with :n:`P`.  (However, the converse doesn't

Reviewer: Except if `P` was already a hypothesis in the context in which case the goal is solved completely.

Author: I reworded the entire description, WDYT?

Reviewer: LGTM now.

>    work, i.e. if the type is :n:`P` and the goal is :n:`~P`.)

Reviewer: ```suggestion
   work: if the type is not a negation and not an empty inductive type, the tactic will fail with "Not a contradiction".)
```

>    .. todo PR: examples?  Or maybe it's simple enough that they're not needed

Reviewer: Sure, examples are always a good thing to provide. However, I would stay mostly with very simple examples (`contradiction` with no arguments) given how limited the usefulness of the variant with arguments is. The only example of `contradiction` with arguments that might be worth to provide is the following (adapted from the stdlib):

```coq
Require Import Arith.

Goal forall (A : Prop), 0 < 0 -> A.
intros. (* In this case, the call to intros is necessary *)
contradiction (Nat.lt_irrefl 0).
Qed.
```

Some simple examples with no argument:

```coq
Inductive F :=. (* Another empty inductive type *)

Goal F -> False.
contradiction.
Qed.
```

```coq
Goal forall (A : Prop), A -> ~A -> False.
contradiction.
Qed.
```

```coq
Goal forall (A : Type) (x : A), ~(x = x) -> False.
contradiction.
Qed.
```

Author: The `intros` is necessary in the first example (`Goal forall (A : Prop), 0 < 0 -> A.`) because the argument to the tactic must be `~P` to match `P`.  Seems to me that tactic would be much simpler for users if the argument could also be `P` to match `~P`.  I can file an issue.  WDYT?

   .. exn:: No such assumption.

>       :undocumented:

Reviewer: Remove this one (it is already documented elsewhere in the chapter) and add "Not a contradiction" instead.

.. tacn:: contradict @ident

   This tactic allows manipulating negated hypothesis and goals. The name
   :n:`@ident` should correspond to a hypothesis. With :n:`contradict H`, the
   current goal and context is transformed in the following way:

>    .. todo PR: what is this below with ¬ and ⊢?  Wrong characters?

Reviewer: These are just Unicode representations of negation (the non-Unicode variant in Coq is `~`, but `¬` is more universal in math / logic) and of a context + goal (the left-hand side is the context and the right-hand side is the goal, exactly as if it was a horizontal line in Coq). Readers with any logical background (even minimal) should be able to understand this.

Author: IIUC the standard meaning of A ⊢ B is "B is provable from A".  In Coq, the meaning is more "We hope to prove B from A, but we don't yet know if that's possible."

I'd rather not introduce new notation that only seems to used in this one spot, but it would take much more space to write these on multiple lines.  `cic.rst` uses turnstile a lot, I think with the second meaning.

Reviewer 3: Indeed. Note that `spéciales` is allows you instantiating parameters not in order if you use `with`.

Reviewer 2: > IIUC the standard meaning of A ⊢ B is "B is provable from A". In Coq, the meaning is more "We hope to prove B from A, but we don't yet know if that's possible."

These are not different things, when considered from the perspective of a logician / proof theorist.  `A ⊢ B` is an assertion that can itself be true or false (or proven or refuted).

Author: I think turnstile is used with varying meanings.  See [here](https://en.wikipedia.org/wiki/Turnstile_(symbol)#:~:text=In%20Unicode%2C%20the%20turnstile%20symbol,assertion%20sign%20(%E2%8A%A6).)).  In several cases it states something that is true rather than being a proposition that can be true or false.  But we don't have to agree on this since I say how to interpret it here (i.e. syntactically).

   impossible. In most cases, :g:`P` or :g:`∼P` is one of the hypotheses of the
   local context.

.. tacn:: contradiction {? @one_term_with_bindings }

   Tries to prove the current goal by finding a contradiction.

   If :n:`@one_term_with_bindings` is not provided (the most common use case),
   the tactic first does an :tacn:`intros`.  The tactic then proves the goal if

   - the updated context has a pair of hypotheses where one is the negation of
     the other (e.g. :n:`P` and not :n:`~P`), or
   - there is a hypothesis with an empty inductive hypothesis (e.g. :n:`False`), or
   - there is a hypothesis that is the negation of a singleton inductive type
     (e.g. :n:`True` or :n:`x=x`).

>    If :n:`@one_term_with_bindings` is provided, the type of
>    :n:`@one_term_with_bindings` must be a negation, such as :n:`~P`,
>    or :n:`False` (or another empty inductive type).

Reviewer: While this sentence is correct, every time I read it, I struggle. I propose to change it like the following, as I think this would make it more symmetric and easier to read:
```suggestion
   If :n:`@one_term_with_bindings` is provided, the type of
   :n:`@one_term_with_bindings` must be either a negation, such as :n:`~P`,
   or an empty inductive type, such as :n:`False`.
```

   If the type is a negation and :n:`P` is a hypothesis in the context,
   the goal is proven.  If the type is a negation and :n:`P` is not in

>    the context, the goal is replaced with :n:`P`.  If the type is :n:`False`,

Reviewer: ```suggestion
   the context, the goal is replaced with :n:`P`.  If the type is :n:`False` or another empty inductive type,
```

   the goal is proven.  Otherwise the tactic fails.  (In particular, note
   that the converse doesn't work, e.g. if the type is :n:`P`, the tactic

>    won't find :n:`~P`.  Use :tacn:`contradict` if you want this behavior.)

Reviewer: Actually, there are two "converse" cases here:

- if there is a hypothesis `P` and the user wants to replace the goal with `~P`, then `contradict` can be used instead
- if there are both hypotheses `H1 : P` and `H2 : ~P` in the context, the user can call `contradiction` without arguments or `contradiction H2`, but `contradiction H1` won't work (and `contradict` is irrelevant to this case).

Author: Perhaps the tactic should be extended to work for your H1/H2 case.  Should I file a PR?

Reviewer: I think I've already expressed my opinion elsewhere that I think it's not worth changing. IMHO the most interesting use of `contradiction` is without arguments anyway.

   impossible. In most cases, :g:`P` or :g:`∼P` is one of the hypotheses of the
   local context.

.. tacn:: contradiction {? @one_term_with_bindings }

   Tries to prove the current goal by finding a contradiction.

   If :n:`@one_term_with_bindings` is not provided (the most common use case),
   the tactic first does an :tacn:`intros`.  The tactic then proves the goal if

   - the updated context has a pair of hypotheses where one is the negation of
     the other (e.g. :n:`P` and not :n:`~P`), or
   - there is a hypothesis with an empty inductive type (e.g. :n:`False`), or
   - there is a hypothesis :n:`~P` where `P` is a singleton inductive type

>      (e.g. :n:`True` or :n:`x=x`) provable by `Goal P. constructor.`.

Reviewer: A bit odd to see two periods IMHO:

![image](https://user-images.githubusercontent.com/1108325/209586881-29c7eda7-ddf1-427c-99ec-be6cd2613d9f.png)

```suggestion
     (e.g. :n:`True` or :n:`x=x`) provable by `Goal P. constructor.`
```

   If :n:`@one_term_with_bindings` is provided, its type
   must be a negation, such as :n:`~P`,
   or an empty inductive type, such as :n:`False`.
   If the type is a negation and :n:`P` is a hypothesis in the context,
   the goal is proven.  If the type is a negation and :n:`P` is not in
   the context, the goal is replaced with :n:`P`.  If the type is :n:`False`
   or another empty inductive type, the goal is proven.
   Otherwise the tactic fails.  (If there is a hypothesis
   `P` and you want to replace the goal with `~P`, use the :tacn:`contradict`
   tactic.  If there are hypotheses `H1 : P` and `H2 : ~P`, use `contradiction`

>    without arguments or `contradiction H2`; `contradiction H1` won't work.)

Reviewer: I suggest replacing ";" with a word, otherwise it looks like what is written is:

>or `contradiction H2; contradiction H1` won't work

![image](https://user-images.githubusercontent.com/1108325/209586937-f20156f4-ac46-4d34-a701-c000ed7fc301.png)

```suggestion
   without arguments or `contradiction H2` since `contradiction H1` won't work.)
```

   impossible. In most cases, :g:`P` or :g:`∼P` is one of the hypotheses of the
   local context.

.. tacn:: contradiction {? @one_term_with_bindings }

   Tries to prove the current goal by finding a contradiction.

>    If :n:`@one_term_with_bindings` is not provided, the tactic first does an

Reviewer: ```suggestion
   If :n:`@one_term_with_bindings` is not provided (most common use case), the tactic first does an
```

   :tacn:`intros`.  The tactic then proves the goal if the updated context has
   a pair of hypotheses where one is the negation of the other (e.g. :n:`P` and
   not :n:`~P`) or if there is a hypothesis with the type :n:`False` (or another

>    empty inductive type).

Reviewer: It looks like we lost the third option "the negation of a singleton inductive type (e.g. True or x=x)" from the previous doc in this paragraph.

Author: OK.  However `1=0` appears to be a singleton inductive and `~ (1 = 0)` is a true and provable statement.  Can you explain?

Reviewer: That's a good point. The previous description was imprecise, but I don't really know how to fix it. Jason's proposal in https://github.com/coq/coq/pull/16659#discussion_r1039512423 was more correct than the previous doc. From your answer to this comment, you seem to think you have some better way of expressing this, but I cannot check it because you forgot to push your changes.

Author: Regarding `(1=0)`, see [here](https://github.com/coq/coq/pull/16659#discussion_r1055977206) and respond there.

   If :n:`@one_term_with_bindings` is provided: The type of
   :n:`@one_term_with_bindings` must be a negation, such as :n:`~P`,
   or :n:`False` (or another empty inductive type).
   If the type is a negation and :n:`P` is a hypothesis in the context,
   the goal is proven.  If the type is a negation and :n:`P` is not in
   the context, the goal is replaced with :n:`P`.  If the type is :n:`False`,
   the goal is proven.  Otherwise the tactic fails.  (In particular, note
   that the converse doesn't work, e.g. if the type is :n:`P`, the tactic

>    won't find :n:`~P`.)

Reviewer: ```suggestion
   won't find :n:`~P`. Use :tacn:`contradict` if you want this behavior.)
```

   Use the :tacn:`discriminate` tactic to prove the current goal when there
   is a hypothesis with an impossible structural equality such as
   :n:`0 = 1`.

.. example:: :tacn:`contradiction` tactic

   Simple examples.  To see more detail, add `intros` after each `Goal`.

   .. coqtop:: reset in

      Inductive F :=. (* Another empty inductive type *)

      Goal F -> False.
      contradiction.
      Qed.

      Goal forall (A : Prop), A -> ~A -> False.
      contradiction.
      Qed.

      Goal forall (A : Type) (x : A), ~(x = x) -> False.
      contradiction.
      Qed.

   Apply a fact from the standard library:

   .. coqtop:: in

      Require Import Arith.

      Goal forall (A : Prop), 0 < 0 -> A.

   .. coqtop:: all

      intros. (* the intros is needed because 0 < 0 is not a negation

>                  while the type of (Nat.lt_ irrefl 0) is a negation (~ 0 < 0) *)

Reviewer: The explanation from the comment is not the good one. It's just that `contradiction` never does an `intros` when used with an argument.

      contradiction (Nat.lt_irrefl 0).
      Qed.

.. tacn:: contradict @ident

>    Transforms the specified hypothesis `:n:`@ident` and the goal in order to

Reviewer: ```suggestion
   Transforms the specified hypothesis :n:`@ident` and the goal in order to
```

   prove that the hypothesis is false. For :n:`contradict H`, the
   current goal and context are transformed as shown.  (For brevity,
   `⊢` is used to separate hypotheses from the goal; it is equivalent to the
   dividing line shown in a context.):

   + H: ~A ⊢ B becomes ⊢ A
   + H: ~A ⊢ ~B becomes H: B ⊢ A
   + H: A ⊢ B becomes ⊢ ~A

>    + H: A ⊢ ~B becomes H: B ⊢ ~A

Reviewer: I don't think that the removal of the Unicode notation for negation is making the reading any better, especially given that none of these are in a verbatim context.

Author: It's more consistent.  This is the only place in the doc that uses ¬.

Reviewer: But isn't it also one of the few places that uses ⊢? I would think this symbol is otherwise probably only used in the typing rules section, and in this section the negation would certainly be represented as ¬ and not as ~.
If you really want to keep the ~ (and I really think this does not help readability), then put some quotes so that this is rendered as code.

Author: Turnstile is used a lot in the more mathematical sections and very little elsewhere.  `¬` is used in the doc only once in `coqdoc.rst`.

![image](https://user-images.githubusercontent.com/1253341/209216130-97c5c6a2-a8dd-4ac3-b81b-d0a3e8ac35a4.png)

> put some quotes so that this is rendered as code

Like this?

![image](https://user-images.githubusercontent.com/1253341/209218398-a35d0a2c-d9ed-4f83-9a7b-bff67d9c2d6b.png)

Reviewer: Yes
